Pluggable verifiable clients

edenhill opened the pull request:
This adds support for pluggable VerifiableConsumer and VerifiableProducers test client implementations
allowing third-party clients to be used in-place of the Java client in kafkatests.

A new VerifiableClientMixin class is added and the standard Java Verifiable{Producer,Consumer}
classes have been changed to use it.

While third-party client drivers may be implemented with a complete class based on the Mixin, a simpler
alternative which requries no kafkatest class implementation is available through the VerifiableClientApp class that uses ducktape's global param to specify the client app to use (passed to ducktape through the `--globals <json>` command line argument).

Some existing kafkatest clients for reference:
Go: https://github.com/confluentinc/confluent-kafka-go/tree/master/kafkatest
Python: https://github.com/confluentinc/confluent-kafka-python/tree/master/confluent_kafka/kafkatest
C++: https://github.com/edenhill/librdkafka/blob/0.9.2.x/examples/kafkatest_verifiable_client.cpp
C#/.NET: https://github.com/confluentinc/confluent-kafka-dotnet/tree/master/test/Confluent.Kafka.VerifiableClient

This PR also contains documentation for the simplex JSON-based verifiable\* client protocol.

There are also some minor improvements that makes troubleshooting failing client tests easier.

edenhill committed:
Added pluggable Verifiable client support

edenhill committed:
Added ConsumerState.Started and alive_nodes()

This extra state is needed to know which nodes succesfully started but have
not yet rebalanced.

edenhill committed:
Add test for pluggable verifiable clients

edenhill committed:
verifiable_consumer: improved debug logging

edenhill committed:
Improved kafkatest error messages

edenhill committed:
OffsetValidationTest.test_consumer_failure: increased robustness

Allow two session timeout cycles: one for timing out the killed client
and one for the rebalance operation.

ijuma commented:
Nice!

jpzk commented:
Cool!

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+
+**Logs:**
+During development of kafkatest clients it is generally a good idea to
+enable collection of the client's stdout and stderr logs for troubleshoot.
```
nit: troubleshoot -> troubleshooting

ewencp replied:
Also, not sure if you want to explain at all why we wouldn't have this on by default. Probably obvious once you start collecting them and see how large they can get, but might not be obvious.

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+
+
+**Deployment:**
+There's currently no automatic way of deploying 3rd party kafkatest clients
```
Are you getting at the fact that the vagrant provisioning scripts don't provide an easy way to make them extensible? Should we consider adding some control in the `Vagrantfile` to include extra provisioning scripts dynamically (i.e. via a config in the `Vagrantfile.local` file)? Over time we've been debating how much of this to assume vs maybe adding some hooks in ducktape to allow you to bake together expected node setup scripts w/ the tests themselves (which is the only way to make them truly cross-cluster type instead of tied directly to Vagrant), but some extra hooks in the Vagrantfile would at least formalize it for the Vagrant cluster types.

edenhill replied:
I don't think we want to poison the vanilla Vagrantfiles, or provisioning scripts, with 3rd-party client specific dependencies.
It is probably better to dockerize the clients, or use some similarily contained.

ewencp replied:
@edenhill I just meant providing a hook (like Vagrantfile.local does for overriding some configs). It wouldn't require much in the Vagrantfile and might be a bit clearer since you wouldn't, e.g., need to know about shared `/vagrant` directories (which in AWS need extra syncing, etc).

But we can always leave this as a future improvement.

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+Then use /vagrant/..../yourkafkatestclient as your run-time path since it will
+now be available on all instances.
+
+The VerifiableClientApp automates the per-worker deployment with the optional
```
Ah, I see. So this will work ok for clients but is also, obviously, highly specific to this class. Maybe this is fine for the current goals since we don't have any other type of clients besides console producer/consumer (which I'm not sure are worth using for these types of client library tests)?

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+    (module_name, clname) = clname.rsplit('.', 1)
+    cluster_mod = importlib.import_module(module_name)
+    impl_class = getattr(cluster_mod, clname)
+    return impl_class(parent, obj)
```
Should we log what class we end up using? Seems like for people trying to get a setup working with another client this will be helpful in making sure they got the configs & ducktape invocation correct.

edenhill replied:
This is logged by impl() in the mixin

edenhill replied:
> `[DEBUG - 2017-03-19 18:53:18,732 - verifiable_client - impl - lineno:194]: Using client implementation VerifiableClientApp for VerifiableProducer`

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+
+    def exec_cmd (self, node):
+        """ :return: command to execute to start instance """
+        return 'echo -e \'{"name": "shutdown_complete" }\n\' ; echo ARGS:'
```
Do we only need the shutdown because none of the tests actually look for the startup message?

ewencp commented on tests/kafkatest/services/verifiable_client.py:
```diff
@@ -0,0 +1,311 @@
+            return []
+
+    def deploy (self, node):
+        """ Call deploy script specified by "deploy" --global key
```
Yeah, so this can work but running it for every test execution kind of sucks since it should generally be a one-time setup. Probably makes sense to eventually factor this out into some sort of setup step that you can run separately and only once.

ewencp replied:
Actually, thinking about this more, we can probably track this per-node and just skip it automatically if it's already been executed on the node. Since this is already a wrapper around the user's deploy script, shouldn't be tough to add in a future patch.

ewencp commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -48,19 +50,27 @@ def handle_shutdown_complete(self):
+    def handle_offsets_committed(self, event, node, logger):
         if event["success"]:
             for offset_commit in event["offsets"]:
+                if "error" in offset_commit and offset_commit["error"] != "":
```
python style nit: You can also shorten tests like this to `if offset_commit.get('error')`. `get` defaults to `None` whereas the index operator will throw an exception if the key doesn't exist.

ewencp commented on tests/kafkatest/services/verifiable_consumer.py:
```diff
@@ -48,19 +50,27 @@ def handle_shutdown_complete(self):
         if event["success"]:
             for offset_commit in event["offsets"]:
+                if "error" in offset_commit and offset_commit["error"] != "":
+                    logger.warn("%s: Offset commit failed for: %s" % (str(node.account), offset_commit))
```
Is the error message useful to log here?

edenhill replied:
failing offset commits leads to later problems, so it is valuable to see this one. Changed it to debug though

ewencp commented on tests/kafkatest/tests/client/pluggable_test.py:
```diff
@@ -0,0 +1,51 @@
+        """
+        consumer = self.setup_consumer(self.TOPIC)
+
+        for num_started, node in enumerate(consumer.nodes, 1):
```
nit: `num_started` from `enumerate` is not used

ewencp commented on tests/kafkatest/tests/client/pluggable_test.py:
```diff
@@ -0,0 +1,51 @@
+
+from kafkatest.tests.verifiable_consumer_test import VerifiableConsumerTest
+
+class PluggableConsumerTest(VerifiableConsumerTest):
```
So, it looks like this doesn't directly verify a pluggable version (e.g. the dummy one which is actually guaranteed to be there). Is this intentional? I'm guessing that means this is only useful when trying to sanity check your client implementation settings and isn't actually useful on, e.g., a nightly basis? I wonder if we should `@ignore` it (or add some similar annotation). It's probably pretty cheap to run so not a big deal either way, but I'm wondering if we want to have some approach for tests like these that are not helpful nightly but helpful when integrating other clients.

edenhill replied:
Good point, will `@ignore` work?

ewencp replied:
We can do the `@ignore`. I'd just look at the cost. If it's tiny compared to everything else, I'd just ignore it. If not, I'd throw in the `@ignore`.

ewencp commented on tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java:
```diff
@@ -256,6 +256,12 @@ String shutdownString() {
         return toJsonString(data);
     }
 
+    String startupString() {
```
Lol, not sure how we ended up with 2 different ways of generating these outputs in VerifiableProducer vs VerifiableConsumer...

ewencp reviewed (commented):
@edenhill A bunch of small comments but this looks good for the most part. Do we know if this cherry-picks cleanly to `0.10.1` as well? I'm debating whether we'll want to apply it there as well (low risk as it's all system test stuff, potentially useful for folks, including us, that want to validate clients against both trunk and 0.10.1 still, but also not really a critical fix and a non-trivial amount of code that could still break something on that branch).

Re: testing, once we refine the review comments, shall we fire off full Jenkins test runs w/o any third party client overrides to validate nothing broke? (Even better would be if we could get a jenkins setup with one of the clients working, but that seems like it'll be much higher overhead than your manual testing atm.)

edenhill committed:
verifiable_client: allow kill_signal to be specified through globals

edenhill committed:
kafkatest: revert VerifiableProducer's new ssh handling

The new file-based SSH handling does not support
cases where the remote program continously outputs stderr output,
such as when debugging is disabled. The stdout channel
will in some cases hang until the remote process is killed.

I've tried to find the cause of this behaviour, through changing
buffering levels, etc, but to no avail, so this commit reinstates
the previous ssh_capture (which VerifiableConsumer still uses).

edenhill commented:
@ewencp With the tests passing the ball is now in your court. :)

ewencp commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -209,24 +182,16 @@ def start_cmd(self, node, idx):
 
         cmd += " --producer.config %s" % VerifiableProducer.CONFIG_FILE
-        cmd += " 2>> %s 1>> %s &" % (VerifiableProducer.STDERR_CAPTURE, VerifiableProducer.STDOUT_CAPTURE)
+        cmd += " 2>> %s | tee -a %s &" % (VerifiableProducer.STDOUT_CAPTURE, VerifiableProducer.STDOUT_CAPTURE)
```
This seems wrong, we have 2 `STDOUT_CAPTURE` and no `STDERR_CAPTURE`. Why did the first one change to `STDOUT` since we're still redirecting stderr to the file?

(It's also not obvious to me why we changed this to use `tee` but it seems to not affect anything critical here.)

edenhill replied:
I'm actually reverting this commit: 62e043a86565cc9bc485658d6c6d176e9aff620f which has STDOUT, STDOUT.
The original STDOUT, STDOUT was added here e6b343302f3208f7f6e0099fe2a7132ef9eaaafb without any notes, so it might just have been a typo.

It actually helps to have stdout (verifiable client json protocol) and stderr combined in the same log for troubleshooting, but I'm not saying it is right.

So in short: This does not add new behaviour, it reverts to the behaviour pre decemember 2016.

I can change it to stderr, stdout but that will require a full re-testing cycle of all my little clients so it will be a while.

ewencp reviewed (commented):
@edenhill LGTM other than one question in `verifiable_producer.py` re: stdout vs stderr. Once we resolve that one issue, we can commit. The other questions/comments can be followed up on later.

Also started https://jenkins.confluent.io/job/system-test-kafka-branch-builder/791 to double check the Java side of things. All looks good for now.

ewencp commented:
Verified against normal java tests here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/797

LGTM, merging to trunk

edenhill commented:
Thanks @ewencp !
